ci, tests: Ignore failed Pods if their equivalent is available

gdemonet opened the pull request:
**Component(s)**: ci, tests

**Context**:

Since K8s 1.18, we are seeing many occurences of Pods being rejected by kubelet and ending up in `NodeAffinity` state. This breaks the implementation of our `wait_pods_status.sh` script, which expects all scheduled Pods to be `Running`. It also would break our `logs should be accessible` E2E test if it were to run, since this test expects all scheduled Pods to be `Running` and have some logs.

**Summary**:

- Rework the stabilization script to only ensure Pods are not transitioning (if a Pod has failed, but remains in that state, it won't prevent this script from succeeding)
- Adjust the workaround for #2726 in `snapshot_upgrade` to use this new script and still ensure all Pods are `Running` before moving on with the upgrade
- Rework our sanity E2E tests to focus on MetalK8s-provided objects, still ignoring failed Pods if any

**Acceptance criteria**:

- Both pre-merge and post-merge pass
- CI runs are not affected by the `NodeAffinity` flaky anymore

gdemonet committed:
ci: Only wait for pods to stabilize

Instead of waiting for all scheduled pods to be running (which is
already done in E2E tests), the "stabilization" script is now only
checking that no change to Pod objects can be seen over a given period
of time.

This will effectively hide any scheduling issue, especially the
`NodeAffinity` flaky which has been impacting us a lot recently.
See kubernetes/kubernetes#92067 for reference.

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @chengyanjin

* @slaperche-scality

* @thomasdanan

* @lucieleonard

* @gdemonet

* @teddyandrieux

* @nootal

* @monpote

* @tcarmet

* @sayf-eddine-scality

* @cmonfort

* @nicolast

* @carlito-scality

* @alexandre-allard-scality

gdemonet committed:
ci: Adapt workaround for loop devices provisioning

We were relying on the `wait_pods_status.sh` script to check that pods
are running, though this changed to only check they are stable.
We thus rework this "trick" to:
- first, run the right Salt state for provisioning loop devices on all
minions
- then, wait for stabilization, with the new `wait_pods_stable.sh`
script
- finally, use `kubectl wait --for=condition=Ready ...` to ensure all
pods are running before we continue

gdemonet committed:
tests: Rework sanity checks

Instead of checking that all Pods scheduled in the cluster are Running
and include logs, we re-focus our sanity checks to:

- verify we can list pods
- verify we can exec commands into a pod
- verify we can get logs from all containers of a pod
- verify that all expected static pods are running
- verify that all expected Deployments are ready
- verify that all expected DaemonSets are ready
- verify that all expected StatefulSets are ready

alexandre-allard reviewed (approved):
LGTM and it seems that you got a green build, lucky guy! :)

gdemonet commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.7`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.7`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`



Please check the status of the associated issue None.

Goodbye gdemonet.
